vtex-apps/order-items
# CHK-463: Batch cart items update operations

#### What problem is this solving?

With the current implementation of our optimistic cart, if you rapidly update an item's quantity (say from 1 to 50), we will make one request for each and every one of those updates (resulting in 50 requests in total), which are enqueued using our `TaskQueue`. Since we are enqueueing the tasks back-to-back, they can take a **looooong** time to complete, delaying some errors that should be shown early to the user (such as not enough stock for the item, or reaching the maximum number of SKUs the store is configured to sell).

#### How to test it?

[Workspace](https://batch--checkoutio.myvtex.com/cart/add?sku=285&qty=1&seller=1&sku=291&qty=1&seller=1&sku=289&qty=1&seller=1)

In the workspace you can see a couple of additional buttons in the cart page (one with the label "adicionar +20" and another with "adicionar +20 (150ms)"), and they do the exact same thing with one small difference.

If you press any of them, they will update the item's quantity one by one, 20 times, so the item will end up with 20 more than it was before. The difference between the two is that the first one will update one by one without any wait time between them, and the second one will wait 150 milliseconds before incrementing.

You can test this PR by pressing the "adicionar +20" button and monitoring the number of requests actually being made using the Network Panel in your browser's DevTools. We shouldn't make the whole 20 requests, and instead batch them together (in this case we will just drop some of the intermediate quantity updates). You can see the previous behavior using [this workspace](https://nobatch--checkoutio.myvtex.com/cart/add?sku=289&qty=21&seller=1).

#### Screenshots or example usage:

Before:

![file](https://user-images.githubusercontent.com/10223856/102623211-eadc5800-4120-11eb-823a-113e0de825f1.gif)

After:

![Dec-18-2020 10-36-27](https://user-images.githubusercontent.com/10223856/102620649-13fae980-411d-11eb-983c-644c3a6a2309.gif)

#### Describe alternatives you've considered, if any.

N/A

#### Related to / Depends on

N/A

#### How does this PR make you feel?

![](https://media.giphy.com/media/5R31AXE8lsERwAEpS6/giphy.gif)
